src/coreclr/src/vm/interoplibinterface.cpp

```diff
@@ -1452,6 +1453,90 @@ bool GlobalComWrappersForMarshalling::TryGetOrCreateObjectForComInstance(
     }
 }
 
+void QCALLTYPE GlobalComWrappersForTrackerSupport::SetGlobalInstanceRegisteredForTrackerSupport()
+{
+    // QCALL contracts are not used here because the managed declaration
+    // uses the SuppressGCTransition attribute
+
+    _ASSERTE(!g_IsGlobalComWrappersRegisteredForTrackerSupport);
+    g_IsGlobalComWrappersRegisteredForTrackerSupport = true;
+}
+
+bool GlobalComWrappersForTrackerSupport::TryGetOrCreateComInterfaceForObject(
+    _In_ OBJECTREF instance,
+    _Outptr_ void** wrapperRaw)
+{
+    if (!g_IsGlobalComWrappersRegisteredForTrackerSupport)
+        return false;
+
+    // Switch to Cooperative mode since object references
+    // are being manipulated.
+    {
+        GCX_COOP();
+
+        // Passing NULL as the ComWrappers implementation indicates using the globally registered instance
+        return TryGetOrCreateComInterfaceForObjectInternal(
+            NULL,
+            instance,
+            CreateComInterfaceFlags::CreateComInterfaceFlags_TrackerSupport,
+            ComWrappersScenario::TrackerSupportGlobalInstance,
+            wrapperRaw);
+    }
+}
+
+bool GlobalComWrappersForTrackerSupport::TryGetOrCreateObjectForComInstance(
+    _In_ IUnknown* externalComObject,
+    _Out_ OBJECTREF* objRef)
+{
+    if (!g_IsGlobalComWrappersRegisteredForTrackerSupport)
+        return false;
+
+    // Determine the true identity of the object
+    SafeComHolder<IUnknown> identity;
+    {
+        GCX_PREEMP();
+
+        HRESULT hr = externalComObject->QueryInterface(IID_IUnknown, &identity);
+        _ASSERTE(hr == S_OK);
+    }
+
+    // Switch to Cooperative mode since object references
+    // are being manipulated.
+    {
+        GCX_COOP();
+
+        // Passing NULL as the ComWrappers implementation indicates using the globally registered instance
+        return TryGetOrCreateObjectForComInstanceInternal(
+            NULL /*comWrappersImpl*/,
+            identity,
+            CreateObjectFlags::CreateObjectFlags_TrackerObject,
+            ComWrappersScenario::TrackerSupportGlobalInstance,
+            NULL /*wrapperMaybe*/,
+            objRef);
+    }
+}
+
+IUnknown* ComWrappersNative::GetIdentityForObject(_In_ OBJECTREF* objectPROTECTED)
+{
+    CONTRACTL
+    {
+        THROWS;
```

Comment:
Since this just trying to get the existing IUnknown, it has no reason to throw or trigger GC.

Reply:
(The fact that it throws and triggers GC is a sign of a performance bug.)

Reply:
I copied the contract from the caller. It's likely allows more than it needs to. I think I can change it to be more restrictive, but I'll try out changing it to nothrow and gc-notrigger locally to validate it.